### DISCUSSION TITLE
Any plan to make `Conv` and `DeConv` use the `Const` architecture ?

### DISCUSSION THREAD

**Discussion by alexander-camuto** [sonos/tract#1206] [General]
Heya :) 

If I recall the tract API recently(ish) made the transition from an architecture where the constantts involved in an operation are carried at the `Op` level, for instance: 

```rust
Matmul {
a: Option<Tensor<i128>>
b: Option<Tensor<i128>>
}
```

to one where constants are represented as independent nodes that are then ingested by the op. 

```rust
Const{
name: "a"
value: Tensor<i128>
}
Const{
name: "b"
value: Tensor<i128>
}
Matmul 
```

A notable exception to this are the `Conv` and `Deconv` nodes which still carry the kernel and bias with them. I was wondering if there way any plan to change this and move these two ops more towards the current API architecture. 

I ask because we have some processing steps that operate on all constants but then have to apply special rules to `Conv` and `DeConv` nodes -- just wondering if this will change.

**Comment by kali** [comment]
Maybe... I am moving away from attributes and towards more inputs, a bit like ONNX has been doing. I can see another benefits to this approach in the Dynamic Convolution models (not sure if these are still around of if the hype has gone) where you compute the weights to the convolution as part of your network. But I don't have definite plan for that.

But I'll keep it in mind, who knows, maybe when I'm masochist enough to go near the convolution code again, I may do it :)

**Reply by alexander-camuto** [reply]
sweet thank you !

**Comment by alexander-camuto** [comment]
Hello @kali recent updates seem to have done this which is amazing. is that correct ?

**Comment by kali** [comment]
Absolutely. And it's been a huge pain, you definitely owe me one if we meet one day :)

**Reply by alexander-camuto** [reply]
very kind -- let me know if you want me to grab you a beer end of feb !
